### PR TITLE
ReindexMembers: check for significant drop in trees and homologies

### PR DESCRIPTION
**Pull request by ens-sb**
## Description
It occured before that there was a considerable loss of gene trees and homologies during reindexing.
This PR add a pipeline step which fails if the loss is significant.

**Related JIRA tickets:**
- ENSCOMPARASW-6397

#### Change 1
Added a runnable for flagging the significant decrease in gene and homology counts.

#### Change 2
Added a pipeline analysis for flagging data decrease.

## Testing
The reindexing pipeline was run on the 111 branch with the murinae collection.

---

For code reviewers: [code review SOP](https://www.ebi.ac.uk/seqdb/confluence/display/EnsCom/Code+review+SOP)

### CODE REVIEW

**Review by Simarpreet-Kaur-Bhurji — APPROVED** [review]

File: modules/Bio/EnsEMBL/Compara/PipeConfig/ReindexMembers_conf.pm [hunk 1]
```diff
@@ -342,6 +342,15 @@ sub core_pipeline_analyses {
             -flow_into      => {
                 '2->A' => 'exon_boundaries_prep',
                 'A->1' => 'pipeline_entry',
+                1      => "hc_deletion",
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/ReindexMembers_conf.pm:345]: Would it be possible to use single quotes here, to stay closer to the style of the rest of the pipe-config file?
```suggestion
                1      => 'hc_deletion',
```

File: modules/Bio/EnsEMBL/Compara/PipeConfig/ReindexMembers_conf.pm [hunk 2]
```diff
@@ -342,6 +342,15 @@ sub core_pipeline_analyses {
...
+            },
+        },
+        {   -logic_name     => 'hc_deletion',
+            -module         => 'Bio::EnsEMBL::Compara::RunnableDB::ReindexMembers::DeletionHealthcheck',
+            -parameters     => {
+                "diff_limit"    => -2.5,
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/ReindexMembers_conf.pm:351]: Also here?
```suggestion
                'diff_limit'    => -2.5,
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    my $self = shift @_;
+
+    $self->param('curr_db', $self);
+
+}
+
+sub run {
+    my $self = shift @_;
+    my $prev_dbc = $self->get_cached_compara_dba('prev_tree_db')->dbc;
+    my $curr_dbc = $self->param('curr_db')->dbc;
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm:56]: If accessing the current pipeline database, the `compara_dba` method could be used instead.
```suggestion
    my $curr_dbc = $self->compara_dba->dbc;
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+use warnings;
+
+use base ('Bio::EnsEMBL::Compara::RunnableDB::BaseRunnable');
+
+
+sub param_defaults {
+
+}
+
+
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm:45]: If no parameter defaults are set here, I'd suggest to remove this.
```suggestion
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+sub fetch_input {
+    my $self = shift @_;
+
+    $self->param('curr_db', $self);
+
+}
+
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm:52]: If you are happy to use the `compara_dba` method below, I'd suggest to remove this.
```suggestion
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    my $diff_limit = $self->param('diff_limit');
+    
+    my $tree_count = $self->count_gene_trees($curr_dbc);
+    my $tree_count_prev = $self->count_gene_trees($prev_dbc);
+    # Calculate the difference in gene tree counts and normalise by the previous count:
+    my $tree_diff = (($tree_count - $tree_count_prev) * 100) / $tree_count_prev;
+    if ($tree_diff < $diff_limit) {
+        my $msg = sprintf("WARNING: The decrease in number of trees is higher than the limit: %.3f%% Current count: %d Previous count: %d!\n", abs($tree_diff), $tree_count, $tree_count_prev);
+        print $msg;
+        $self->throw($msg)
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm:66]: If creating a new Perl Hive runnable with a non-transient error, I'd suggest to use `die_no_retry`:
```suggestion
        $self->die_no_retry($msg)
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    my $hom_count = $self->count_homologies($curr_dbc);
+    my $hom_count_prev = $self->count_homologies($prev_dbc);
+    # Calculate the difference in homology counts and normalise by the previous count:
+    my $hom_diff = (($hom_count - $hom_count_prev) * 100) / $hom_count_prev;
+    if ($hom_diff < $diff_limit) {
+        my $msg = sprintf("WARNING: The decrease in number of homologies is higher than the limit: %.3f%% Current count: %d Previous count: %d!\n", abs($hom_diff), $hom_count, $hom_count_prev);
+        print $msg;
+        $self->throw($msg)
```
Comment (twalsh-ebi) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ReindexMembers/DeletionHealthcheck.pm:76]:
```suggestion
        $self->die_no_retry($msg)
```

**Review by twalsh-ebi — APPROVED** [review]

**Pull request merged**